bundles/config/org.eclipse.smarthome.config.xml/src/main/java/org/eclipse/smarthome/config/xml/AbstractXmlBasedProvider.java

```diff
@@ -0,0 +1,259 @@
+/**
+ * Copyright (c) 2014-2017 by the respective copyright holders.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.eclipse.smarthome.config.xml;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+import org.osgi.framework.Bundle;
+
+/**
+ * Common base class for XML based providers.
+ *
+ * @author Simon Kaufmann - initial contribution, factored out of subclasses
+ *
+ * @param <T_ID> the key type, e.g. ThingTypeUID, ChannelUID, URI,...
+ * @param <T_OBJECT> the object type, e.g. ThingType, ChannelType, ConfigDescription,...
+ */
+public abstract class AbstractXmlBasedProvider<T_ID, T_OBJECT> {
+
+    private class LocalizedKey {
+        public T_ID uid;
```

Comment:
As this is a generic and also an URI could be an ID (refer to your JavaDoc) we should name the variable `id` instead of `uid`.